<a href="https://colab.research.google.com/github/ValentinBarco/Analog-Mixer-desing/blob/main/Manejo%20de%20datasets/Manejo_basico_de_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

## Creacion de df


In [ ]:
# Creo atributos de ventas de dispositivos electronicos
producto= ['Laptop', 'Teléfono', 'Televisor', 'Tablet',  'Cámara', 'Smartwatch', 'Auriculares']
precio= [800, 500, 1000, 300, 400, 200, 150]
unidades_vendidas = [50, 120, 80, 30, 65, 70, 40]
total_Ventas= [40000, 60000, 80000, 9000, 18000, 18000, 6000]

# Creao el DataFrame con el nombre de cada columna y los valores que debe tener cada columna
df = pd.DataFrame({'Producto': producto, 'Precio': precio,'Unidades': unidades_vendidas})

df.describe() #Muestra características del dataset (cantidad, media, desvio, min, 25%, 50%, 75% y max)
df.tail() #Muestra los ULTIMOS 5, también se puede aplicar .tail(10) y muestra los ultimos 10
df.head() #Muestra los PRIMEROS 5, también se puede aplicar .head(10) y muestra los primeros 10
df #Muestra un paneo general del dataset

Producto  Precio  Unidades
0     Laptop     800        50
1   Teléfono     500       120
2  Televisor    1000        80
3     Tablet     300        30
4     Cámara     400        65

In [ ]:
# Elijo una columna que quiera visualizar
df['Producto']

0         Laptop
1       Teléfono
2      Televisor
3         Tablet
4         Cámara
5     Smartwatch
6    Auriculares
Name: Producto, dtype: object

In [ ]:
# Elijo las columnas Producto y Unidades, todas las filas
df_filtrado = df.loc[:,['Producto','Unidades']]

df_filtrado

Producto  Unidades
0       Laptop        50
1     Teléfono       120
2    Televisor        80
3       Tablet        30
4       Cámara        65
5   Smartwatch        70
6  Auriculares        40

In [ ]:
# Elijo las columnas Producto y Unidades de la fila 0 a la 2
df_filtrado = df.loc[0:2,['Producto','Unidades']]

df_filtrado

Producto  Unidades
0     Laptop        50
1   Teléfono       120
2  Televisor        80

In [ ]:
# Elijo la primer fila
df_producto_solo = df.loc[0]

df_producto_solo

Producto    Laptop
Precio         800
Unidades        50
Name: 0, dtype: object

In [ ]:
# Elijo de la fila 0 a la 3
df_producto_varios = df.loc[0:3]

df_producto_varios

Producto  Precio  Unidades
0     Laptop     800        50
1   Teléfono     500       120
2  Televisor    1000        80
3     Tablet     300        30

In [ ]:
# Calculo las ventas totales
df['Ventas_totales'] = df['Precio'] * df['Unidades']

df_dispositivos = df

df_dispositivos

Producto  Precio  Unidades  Ventas_totales
0       Laptop     800        50           40000
1     Teléfono     500       120           60000
2    Televisor    1000        80           80000
3       Tablet     300        30            9000
4       Cámara     400        65           26000
5   Smartwatch     200        70           14000
6  Auriculares     150        40            6000

In [ ]:

# Creo los datos a trabajar
producto = ['Monitor', 'Teclado', 'Impresora', 'Mouse', 'Disco Duro', 'Tarjeta gráfica', 'Memoria RAM', 'Procesador']
unidades_vendidas = [70, 200, 100, 50, 35, 80, 60, 40]
precio = [400, 150, 800, 25, 100, 300, 120, 350]

# Otra manera de agrupar los datos y crear un DataFrame
data = {

    'Producto': producto,
    'Precio': precio,
    'Unidades': unidades_vendidas,
}

df_componentes = pd.DataFrame(data)


# Calculo las ventas totales del nuevo dataset
df_componentes['Ventas_totales'] = df_componentes['Precio'] * df_componentes['Unidades']

df_componentes




Producto  Precio  Unidades  Ventas_totales
0          Monitor     400        70           28000
1          Teclado     150       200           30000
2        Impresora     800       100           80000
3            Mouse      25        50            1250
4       Disco Duro     100        35            3500
5  Tarjeta gráfica     300        80           24000
6      Memoria RAM     120        60            7200
7       Procesador     350        40           14000

## Concatenar df


In [ ]:
# Uno ambos DataFrames
df_concatenado = pd.concat([df_dispositivos, df_componentes], ignore_index=True)

df_concatenado

Producto  Precio  Unidades  Ventas_totales
0            Laptop     800        50           40000
1          Teléfono     500       120           60000
2         Televisor    1000        80           80000
3            Tablet     300        30            9000
4            Cámara     400        65           26000
5        Smartwatch     200        70           14000
6       Auriculares     150        40            6000
7           Monitor     400        70           28000
8           Teclado     150       200           30000
9         Impresora     800       100           80000
10            Mouse      25        50            1250
11       Disco Duro     100        35            3500
12  Tarjeta gráfica     300        80           24000
13      Memoria RAM     120        60            7200
14       Procesador     350        40           14000

## Filtros



In [ ]:
# Filtro productos con mas de 40 unidades vendidas
df_alto_valor = df.loc[df['Unidades'] > 40]

df_alto_valor

Producto  Precio  Unidades  Ventas_totales
0      Laptop     800        50           40000
1    Teléfono     500       120           60000
2   Televisor    1000        80           80000
4      Cámara     400        65           26000
5  Smartwatch     200        70           14000

In [ ]:
# Elijo los productos que deseo quedarme
productos_seleccionados = ['Procesador', 'Memoria RAM', 'Mouse', 'Monitor', 'ML Engineer']

# Aplico el filtro de los productos seleccionados
df_seleccionados = df_concatenado[df_concatenado['Producto'].isin(productos_seleccionados)]

df_seleccionados

Producto  Precio  Unidades  Ventas_totales
7       Monitor     400        70           28000
10        Mouse      25        50            1250
13  Memoria RAM     120        60            7200
14   Procesador     350        40           14000

## Promedios

In [ ]:
# Realizo un promedio de las unidades vendidas
promedio_unidades = df['Unidades'].mean()

print(f"La cantidad promedio de unidades vendidas es {promedio_unidades}")


La cantidad promedio de unidades vendidas es 65.0


In [ ]:
# Realizo un promedio del precio
promedio_precio = df['Precio'].mean()

print(f"El precio promedio de unidades vendidas es {promedio_precio}")

El precio promedio de unidades vendidas es 478.57142857142856


## Importacion de CSV

In [ ]:
# Cargo un dataset ya existente
ruta_de_acceso = '/content/Dataset_de_ventas_ejemplo_CLASE1.csv'

df_ventas = pd.read_csv(ruta_de_acceso)

df_ventas.head(10)

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1      NaN                         NaN              NaN        NaN   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   
5   176561            Wired Headphones                1      11.99   
6   176562        USB-C Charging Cable                1      11.95   
7   176563  Bose SoundSport Headphones                1      99.99   
8   176564        USB-C Charging Cable                1      11.95   
9   176565          Macbook Pro Laptop                1       1700   

       Order Date                        Purchase Address  
0  04/19/19 08:46            917 1st St, Dallas, TX 75001  
1             NaN                                     NaN  
2  04/07/19 22:30       682 Chestnut St, Boston, MA 02215  
3  04/12/19 14:38    669 Spruce St, Los Angeles, CA 90001  
4  04/12/19 14:38    669 Spruce St, Los Angeles, CA 90001  
5  04/30/19 09:27       333 8th St, Los Angeles, CA 90001  
6  04/29/19 13:03  381 Wilson St, San Francisco, CA 94016  
7  04/02/19 07:46        668 Center St, Seattle, WA 98101  
8  04/12/19 10:58         790 Ridge St, Atlanta, GA 30301  
9  04/24/19 10:38  915 Willow St, San Francisco, CA 94016

In [ ]:
# Eliminamos las filas que contienen NaN
df_ventas = df_ventas.dropna()

# Cambia los Nan por 0
df_ventas = df_ventas.fillna(0)

df_ventas = df_ventas.fillna({'Quantity Ordered':0,'Price Each':2})

df_ventas

Order ID                     Product Quantity Ordered Price Each  \
0       176558        USB-C Charging Cable                2      11.95   
2       176559  Bose SoundSport Headphones                1      99.99   
3       176560                Google Phone                1        600   
4       176560            Wired Headphones                1      11.99   
5       176561            Wired Headphones                1      11.99   
...        ...                         ...              ...        ...   
18378   194090                Google Phone                1        600   
18379   194091       AA Batteries (4-pack)                1       3.84   
18380   194092      AAA Batteries (4-pack)                2       2.99   
18381   194093       AA Batteries (4-pack)                1       3.84   
18382   194094    Lightning Charging Cable                1      14.95   

           Order Date                        Purchase Address  
0      04/19/19 08:46            917 1st St, Dallas, TX 75001  
2      04/07/19 22:30       682 Chestnut St, Boston, MA 02215  
3      04/12/19 14:38    669 Spruce St, Los Angeles, CA 90001  
4      04/12/19 14:38    669 Spruce St, Los Angeles, CA 90001  
5      04/30/19 09:27       333 8th St, Los Angeles, CA 90001  
...               ...                                     ...  
18378  04/08/19 17:11   177 Jackson St, Los Angeles, CA 90001  
18379  04/15/19 16:02         311 Forest St, Austin, TX 73301  
18380  04/28/19 14:36  347 Sunset St, San Francisco, CA 94016  
18381  04/14/19 15:09         835 Lake St, Portland, OR 97035  
18382  04/18/19 11:08          354 North St, Boston, MA 02215  

[18324 rows x 6 columns]

In [ ]:
# Transformo todos los valores no numericos en Nan en las columnas Quantity Ordered y Price Each
df_ventas['Quantity Ordered'] = pd.to_numeric(df_ventas['Quantity Ordered'], errors='coerce')  #El errors= 'coerce' transforma los valores no numericos a NaN'
df_ventas['Price Each'] = pd.to_numeric(df_ventas['Price Each'], errors='coerce')

# Elimino todos los Nan y me quedo solo con los valores numericos
df_ventas = df_ventas.dropna()

# Transformo todos los valores numericos al tipo de dato int
df_ventas['Quantity Ordered'] = df_ventas['Quantity Ordered'].astype(int)
df_ventas['Price Each'] = df_ventas['Price Each'].astype(int)

In [ ]:
# Un ejemplo de fiiltrado por valor: En este caso me quedo con los productos que se filtraron
df_ventas_productos_mas_vendidos = df_ventas.loc[df_ventas['Quantity Ordered'] > 3]

# Muestro los primeros 5
df_ventas_productos_mas_vendidos.head()

Order ID                 Product  Quantity Ordered  Price Each  \
480    177017  AAA Batteries (4-pack)                 5           2   
530    177065  AAA Batteries (4-pack)                 4           2   
628    177160  AAA Batteries (4-pack)                 7           2   
1259   177757   AA Batteries (4-pack)                 4           3   
1348   177838  AAA Batteries (4-pack)                 6           2   

          Order Date                        Purchase Address  
480   04/16/19 00:04  713 Sunset St, New York City, NY 10001  
530   04/03/19 21:04        171 Madison St, Boston, MA 02215  
628   04/26/19 11:37            689 7th St, Dallas, TX 75001  
1259  04/25/19 11:03  879 Walnut St, New York City, NY 10001  
1348  04/29/19 00:16    872 Pine St, San Francisco, CA 94016

### **PREGUNTA:  ¿Qué producto se vendió más?**

## Agrupaciones

In [ ]:
# Agrupo los productos y la cantidad ordenada y realizo la sumatoria de los productos vendidos
productos_vendidos = df_ventas.groupby('Product')['Quantity Ordered'].sum() #Aca se puede hacer muchas operaciones .sum() .min() .max() .std() etc

# Ordeno los de mayor a menor segun la cantidad de productos vendidos
productos_vendidos_ordenados = productos_vendidos.sort_values(ascending=False)

productos_vendidos_ordenados

Product
AAA Batteries (4-pack)        2936
AA Batteries (4-pack)         2822
Lightning Charging Cable      2371
USB-C Charging Cable          2270
Wired Headphones              2066
Apple Airpods Headphones      1525
Bose SoundSport Headphones    1288
27in FHD Monitor               737
iPhone                         695
34in Ultrawide Monitor         653
Google Phone                   581
27in 4K Gaming Monitor         565
Flatscreen TV                  459
Macbook Pro Laptop             455
20in Monitor                   395
ThinkPad Laptop                392
Vareebadd Phone                220
LG Dryer                        77
LG Washing Machine              61
Name: Quantity Ordered, dtype: int64

El producto que mas se vendio fue AAA Batteries (4-pack) con 2936 unidades vendidas

### **PREGUNTA: ¿Que producto tuvo mas ventas totales?**

## Agrupamiento múltiple con calculos

In [ ]:
# Realizo un agrupamiento de los datos donde a la cantidad ordenada le aplico la suma, a el precio el promedio y lo agrupo con los productos.
Analisis_de_productos = df_ventas.groupby('Product').agg({
    'Quantity Ordered': 'sum',
    'Price Each': 'mean'
})

# Ordeno de mayor a menor segun la cantidad ordenada
Analisis_de_productos_ordenados = Analisis_de_productos.sort_values(by='Quantity Ordered',ascending=False)

Analisis_de_productos_ordenados

Quantity Ordered  Price Each
Product                                                 
AAA Batteries (4-pack)                  2936         2.0
AA Batteries (4-pack)                   2822         3.0
Lightning Charging Cable                2371        14.0
USB-C Charging Cable                    2270        11.0
Wired Headphones                        2066        11.0
Apple Airpods Headphones                1525       150.0
Bose SoundSport Headphones              1288        99.0
27in FHD Monitor                         737       149.0
iPhone                                   695       700.0
34in Ultrawide Monitor                   653       379.0
Google Phone                             581       600.0
27in 4K Gaming Monitor                   565       389.0
Flatscreen TV                            459       300.0
Macbook Pro Laptop                       455      1700.0
20in Monitor                             395       109.0
ThinkPad Laptop                          392       999.0
Vareebadd Phone                          220       400.0
LG Dryer                                  77       600.0
LG Washing Machine                        61       600.0

In [ ]:
# Calculo las ventas totales como la cantidad de unidades vendidas por el precio
Analisis_de_productos_ordenados['Total Sales'] = Analisis_de_productos_ordenados['Quantity Ordered'] * Analisis_de_productos_ordenados['Price Each']

Analisis_de_productos_ordenados

Quantity Ordered  Price Each  Total Sales
Product                                                              
AAA Batteries (4-pack)                  2936         2.0       5872.0
AA Batteries (4-pack)                   2822         3.0       8466.0
Lightning Charging Cable                2371        14.0      33194.0
USB-C Charging Cable                    2270        11.0      24970.0
Wired Headphones                        2066        11.0      22726.0
Apple Airpods Headphones                1525       150.0     228750.0
Bose SoundSport Headphones              1288        99.0     127512.0
27in FHD Monitor                         737       149.0     109813.0
iPhone                                   695       700.0     486500.0
34in Ultrawide Monitor                   653       379.0     247487.0
Google Phone                             581       600.0     348600.0
27in 4K Gaming Monitor                   565       389.0     219785.0
Flatscreen TV                            459       300.0     137700.0
Macbook Pro Laptop                       455      1700.0     773500.0
20in Monitor                             395       109.0      43055.0
ThinkPad Laptop                          392       999.0     391608.0
Vareebadd Phone                          220       400.0      88000.0
LG Dryer                                  77       600.0      46200.0
LG Washing Machine                        61       600.0      36600.0

In [ ]:
# Ordeno de mayor a menor segun las ventas totales
Analisis_de_productos_ordenados = Analisis_de_productos.sort_values(by='Total Sales',ascending=False)

Analisis_de_productos_ordenados

Quantity Ordered  Price Each  Total Sales
Product                                                              
Macbook Pro Laptop                       455      1700.0     773500.0
iPhone                                   695       700.0     486500.0
ThinkPad Laptop                          392       999.0     391608.0
Google Phone                             581       600.0     348600.0
34in Ultrawide Monitor                   653       379.0     247487.0
Apple Airpods Headphones                1525       150.0     228750.0
27in 4K Gaming Monitor                   565       389.0     219785.0
Flatscreen TV                            459       300.0     137700.0
Bose SoundSport Headphones              1288        99.0     127512.0
27in FHD Monitor                         737       149.0     109813.0
Vareebadd Phone                          220       400.0      88000.0
LG Dryer                                  77       600.0      46200.0
20in Monitor                             395       109.0      43055.0
LG Washing Machine                        61       600.0      36600.0
Lightning Charging Cable                2371        14.0      33194.0
USB-C Charging Cable                    2270        11.0      24970.0
Wired Headphones                        2066        11.0      22726.0
AA Batteries (4-pack)                   2822         3.0       8466.0
AAA Batteries (4-pack)                  2936         2.0       5872.0

El producto que tuvo mas ventas totales fue la Macbook Pro Laptop con 773500.0 U$S